flocker/node/_deploy.py

```diff
@@ -497,25 +484,71 @@ def got_volumes(available_manifestations):
                 for (dataset_id, maximum_size) in
                 available_manifestations.values())
 
-            return NodeState(
+            return [NodeState(
                 hostname=self.hostname,
                 applications=None,
                 used_ports=None,
                 manifestations={manifestation.dataset_id: manifestation
                                 for manifestation in manifestations},
                 paths=manifestation_paths,
-            )
+            )]
         volumes.addCallback(got_volumes)
         return volumes
 
-    def calculate_necessary_state_changes(self, *args, **kwargs):
-        # Does nothing in this branch. Follow up will move
-        # calculate_necessary_state_changes code here:
-        # https://clusterhq.atlassian.net/browse/FLOC-1553
-        return Sequentially(changes=[])
+    def calculate_changes(self, configuration, cluster_state):
+        """
+        Calculate changes to peer-to-peer manifestations.
+        """
+        local_state = cluster_state.get_node(self.hostname)
+        # We need to know applications (for now) to see if we should delay
+        # deletion or handoffs.
+        if local_state.applications is None:
+            return Sequentially(changes=[])
+        phases = []
+
+        # For now we delay deletion and handoffs until we know application
+        # isn't using the dataset. Later on we'll use leases to decouple
+        # the application and dataset logic better.
+        in_use_datasets = {app.volume.manifestation.dataset_id
+                           for node in cluster_state.nodes
+                           for app in node.applications
+                           if app.volume is not None}
+
+        # Find any dataset that are moving to or from this node - or
+        # that are being newly created by this new configuration.
+        dataset_changes = find_dataset_changes(
+            self.hostname, cluster_state, configuration)
+
+        if dataset_changes.resizing:
+            phases.append(InParallel(changes=[
+                ResizeDataset(dataset=dataset)
+                for dataset in dataset_changes.resizing]))
+
+        going = [handoff for handoff in dataset_changes.going
+                 if handoff.dataset.dataset_id not in in_use_datasets]
+        if going:
+            phases.append(InParallel(changes=[
+                HandoffDataset(dataset=handoff.dataset,
+                               hostname=handoff.hostname)
+                for handoff in going]))
+
```

Comment:
Ooh.  No need to handling `coming` anymore, I see.  The other side pushes stuff here and ... it shows up when it shows up, no need to do anything special.  Neat.